source/extensions/filters/network/dubbo_proxy/dubbo_protocol_impl.cc

```diff
@@ -103,18 +135,80 @@ bool DubboProtocolImpl::decode(Buffer::Instance& buffer, Protocol::Context* cont
         std::make_unique<RequestMessageImpl>(request_id, body_size, is_event);
     req->fromBuffer(buffer);
     context->is_request_ = true;
-    callbacks_.onRequestMessage(std::move(req));
+    callbacks_->onRequestMessage(std::move(req));
   } else {
     ResponseMessageImplPtr res =
         std::make_unique<ResponseMessageImpl>(request_id, body_size, is_event);
     res->fromBuffer(buffer);
-    callbacks_.onResponseMessage(std::move(res));
+    callbacks_->onResponseMessage(std::move(res));
   }
 
   buffer.drain(MessageSize);
   return true;
 }
 
+bool DubboProtocolImpl::decode(Buffer::Instance& buffer, Protocol::Context* context,
+                               MessageMetadataSharedPtr metadata) {
+  if (!metadata) {
+    throw EnvoyException("invalid metadata parameter");
+  }
+
+  if (buffer.length() < DubboProtocolImpl::MessageSize) {
+    return false;
+  }
+
+  uint16_t magic_number = buffer.peekBEInt<uint16_t>();
+  if (magic_number != MagicNumber) {
+    throw EnvoyException(fmt::format("invalid dubbo message magic number {}", magic_number));
+  }
+
+  uint8_t flag = buffer.peekInt<uint8_t>(FlagOffset);
+  MessageType type =
+      (flag & MessageTypeMask) == MessageTypeMask ? MessageType::Request : MessageType::Response;
+  bool is_event = (flag & EventMask) == EventMask ? true : false;
+  int64_t request_id = buffer.peekBEInt<int64_t>(RequestIDOffset);
+  int32_t body_size = buffer.peekBEInt<int32_t>(BodySizeOffset);
+
+  if (body_size > MaxBodySize || body_size <= 0) {
+    throw EnvoyException(fmt::format("invalid dubbo message size {}", body_size));
+  }
+
+  metadata->setMessageType(type);
+  metadata->setRequestId(request_id);
+
+  if (type == MessageType::Request) {
+    parseRequestInfoFromBuffer(buffer, metadata);
+  } else {
+    parseResponseInfoFromBuffer(buffer, metadata);
+  }
+
+  context->header_size_ = DubboProtocolImpl::MessageSize;
+  context->body_size_ = body_size;
+  context->is_heartbeat_ = is_event;
```

Comment:
Where is the logic for heartbeat message processing?

Reply:
The heartbeat processing logic is in the Decoder and ConnectionManager, which is not included in this submission.